core/camel-core/src/test/java/org/apache/camel/processor/OnCompletionBeforeChainedSedaRoutes.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.processor;
+
+import org.apache.camel.ContextTestSupport;
+import org.apache.camel.builder.RouteBuilder;
+import org.junit.jupiter.api.Test;
+
+public class OnCompletionBeforeChainedSedaRoutes extends ContextTestSupport {
+
+    @Test
+    public void testOnCompletionChained() throws Exception {
+        String body = "<body>";
+
+        getMockEndpoint("mock:end").expectedBodiesReceived(body);
```

Comment:
I'd expect `"mock:end"` to arrive only after `"mock:cdone"`.

Reply:
Due to the fact that the SEDA routes are run asynchronously, the order of arrival is indeterminate in the beforeConsumer mode. Also the order in which the expectedBodiesRecieved method is invoked does not control the order in which the assertions are actually evaluated. 
In the afterConsumer mode, the order of arrival is determinate: end, ddone, cdone, bdone, adone.

Reply:
Indeed 👍🏻
I only later noticed that those are expectations for a set of different endpoints.

I'll write a test that delays processing on the second queue and ensures that the "on completion" gets executed meanwhile.

Reply:
That would be great. It was beyond my vacation-mode motivation. -:)

Reply:
[Here](https://github.com/apache/camel/pull/8120) you go. Looks nice! Thank you for your effort 👍🏻